instrumentserver/dashboard/logger.py

```diff
@@ -0,0 +1,175 @@
+import pandas as pd
+from .. import QtCore
+
+from ..client import Client as InstrumentClient
+
+from .__init__ import read_config
+
+
+class LoggerParameters:
+    """
+    Holds the different parameters the logger is tracking. It holds all of the metadata as well as fresh data
+
+    :param name: Name of the parameter.
+    :param source_type: Specifies how to gather the data for the parameter (parameter or broadcast).
+    :param parameter_path: Full name with submodules of the qcodes parameter.
+    :param server: Location of the server, defaults to 'localhost'.
+    :param port: Port of the server, defaults to 5555.
+    :param interval: Interval of time to gather new updates in seconds,
+                     only impactful if source_type is of the parameter type. defaults to 1.
+    """
+
+    def __init__(self, name: str,
+                 source_type: str,
+                 parameter_path: str,
+                 server: Optional[str] = 'localhost',
+                 port: Optional[int] = 5555,
+                 interval: Optional[int] = 1):
+
+        # load values
+        self.name = name
+        self.source_type = source_type
+        self.parameter_path = parameter_path
+        self.server = server
+        self.port = port
+        self.address = f"tcp://{self.server}:{self.port}"
+        self.interval = interval
+
+        self.data = []
+        self.time = []
+
+        # locate the instrument this parameter is located
+        submodules = parameter_path.split('.')
+        self.instrument_name = submodules[0]
+        self.client = InstrumentClient(self.server, self.port)
+        self.instrument = self.client.get_instrument(self.instrument_name)
+
+        # get the name of the parameter with submodules
+        parameter_name = ''
+        for i in range(1, len(submodules)):
+            parameter_name = parameter_name + submodules[i]
+        self.parameter_name = parameter_name
+
+        # record the time the parameter was created
+        self.last_saved_t = datetime.datetime.now()
+
+    def update(self):
+        """
+        Updates the parameter and save a new data point in memory
+        """
+
+        # check that the source type is parameter
+        if self.source_type == 'parameter':
+
+            # just for development
+            self.instrument.generate_data(self.parameter_name)
+
+            # gather new data and save it in memory
+            new_data = self.instrument.get(self.parameter_name)
+            current_time = datetime.datetime.now()
+            self.data.append(new_data)
+            self.time.append(current_time)
+
+        if self.source_type == 'broadcast':
+            raise NotImplementedError
+
+
+class ParameterLogger(QtCore.QObject):
+    """
+    Main class of the logger. All of the parameters are saved inside this class
+    """
+    def __init__(self):
+        super().__init__()
+
+        # used for testing, the instruments should be already created for the dashboard to work
```

Comment:
i'd prefer if the test wasn't hardcoded into this file. a better solution would be to have a separate testing script.